### ISSUE TITLE
TypeScript styles / norms

### ISSUE THREAD

**Issue by danvk** [refstudio/refstudio#18]
Kicking off a discussion. As a general rule style guidelines should all be automated so that code reviews don't get bogged down by discussion of whitespace preferences.

- Use prettier for code formatting.
  - Everyone should enable "format-on-save"
  - We should check formatting as part of CI
  - Single quotes, two character indent, yes to semicolons, 80 character width (some groups prefer 100)
- Use a recent version of TypeScript with `strict` set
- Use eslint for linting
  - Enable `eslint:recommended`
  - Enable [`@typescript-eslint/recommended'](eslint:recommended', 'plugin:@typescript-eslint/recommended)
  - Enable `plugin:react/recommended` and `plugin:react-hooks/recommended`

Some other suggestions of mine, open to discussion.

General JavaScript:

- Avoid `export default` (named exports encourage consistent use of the same name for the same thing).
- Use a canonical ordering for imports with three (or fewer) groups separated by a blank line, alphabetized within each group (this can be automated with the [`simple-import-sort`](https://github.com/lydell/eslint-plugin-simple-import-sort) eslint plugin):
    ```ts
    import {foo} from 'node-module1';
    import {bar} from 'node-module2';

    import {baz} from './relative-import/a';
    import {quux} from './relative-import/b';

    import './static-asset.css';
    ```
- Use `async` / `await` instead of raw Promises and `.then` whenever possible.

A few TypeScript-specific rules:

- Avoid `any` wherever reasonably possible.
- Prefer `interface` to `type` where possible for simple object types (your type will show up more predictably in error messages and it will improve TypeScript compiler performance for large code bases).
- Prefer `T[]` to `Array<T>` unless `T` is very complex and the latter reads more clearly.
- Avoid non-standard TypeScript language extensions such as `enum` and [parameter properties](https://typescript-eslint.io/rules/parameter-properties/)
- Use synthetic default imports where possible, e.g. `import React from 'react';` instead of `import * as React from 'react';`

@cguedes @sergioramos @sehyod would love feedback and suggestions!

**Comment by cguedes** [comment]
@danvk I agree with everything above. I'm already using almost all of these in my projects. I really like the "format on save" rule.

The only one I would like to change is the character width rule from 80 to *120*.  80 makes sense for prose code writing (ex: typescript) but we will have react and there 120 works best IMO otherwise you can very easily get wrapped to multiple lines.

<img width="1164" alt="image" src="https://github.com/refstudio/refstudio/assets/174127/6ce6371a-88fe-490e-8dc6-60d6388f1f9d">

What are your thoughts on having the `.vscode > settings.json` in the repo? That would allow us to enforce some of the rules you mention like the auto save and formatter (and others like `explorer.excludeGitIgnore`).

<img width="996" alt="image" src="https://github.com/refstudio/refstudio/assets/174127/0c7d074a-fef7-4d42-aa56-15d4985c3a49">

**Comment by danvk** [comment]
We can do 120. I’ve done 100 on previous projects. My main concern is being
able to fit two text files side by side on a reasonably sized monitor.

**Comment by cguedes** [comment]
This was fixed by #24
